### PR TITLE
persist: address a batch of encoding TODOs

### PR DESCRIPTION
**Pull request by danhhz**
See individual commits for details.

### Motivation

   * This PR refactors existing code.

Cleanups in prep for protobuf encoding of BlobMeta.

### Tips for reviewer

I left the unsealed_ts_upper here as a TODO because the rest of this PR is really straightforward refactoring and I wanted to keep it really easy to review. I'll get to it this week.

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [N/A] This PR adds a release note for any
  [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/user/content/release-notes.md#what-changes-require-a-release-note).

### CODE REVIEW

**Comment by ruchirK** [comment]
haven't looked at this in detail, but one small thought here is -- maybe do we want to bump meta/update the golden test on each commit? I guess the two cases I'm thinking about are:

1. In case we want to revert a subset of the commits for some reason it would be a bit simpler
2. In case someone else tries to build and run MZ with a subset of the commits here that doesn't include the final commit

I dunno how compelling either of those reasons are, and I know bumping meta/updating golden each time is tedious. Wasn't sure if doing it each time also came with other downsides.

No strong preference either way, so feel free to leave it as is if you feel strongly

**Comment by danhhz** [comment]
I was 50/50 on this. Tbh, neither of those reasons are particularly compelling to me, but it's easy enough to switch that 51/49 is worth it. I'll have to force push it though, have you started reviewing yet? If so, I'll just do it after the LGTM

**Comment by ruchirK** [comment]
I'll review asap -- gimme like 15 minutes (ive already seen most of these changes so it shouldn't take long)

**Comment by danhhz** [comment]
> 1. In case we want to revert a subset of the commits for some reason it would be a bit simpler

Of the two, this is more compelling. In theory I agree with it. In practice, no one is going to want to revert any of these commits individually. Setting a good precedent is probably where I go from 50/50 to 51/49 on this :)

> 2. In case someone else tries to build and run MZ with a subset of the commits here that doesn't include the final commit

This to me is a non-goal. I have a reasonably strong opinion that the unit of atomic work is a PR, not a commit. I suppose this also negates (1) a bit, but I'm still willing to change it.

File: src/persist/src/indexed/encoding.rs [hunk 1]
```diff
@@ -111,20 +107,39 @@ pub struct StreamRegistration {
...
 /// Invariants:
-/// - The batch SeqNo ranges are sorted and non-overlapping.
+/// - The unsealed_batch SeqNo ranges are sorted and non-overlapping.
+/// - The trace_batch Descriptions are sorted, non-overlapping, and contiguous.
+/// - The since frontier is either 0 or < the trace's sealed frontier.
+/// - Every batch's since frontier is <= the overall trace's since frontier.
+/// - The compaction level of trace_batches is weakly decreasing when iterating
+///   from oldest to most recent time intervals.
+/// - Every trace_batch's upper is <= the overall trace's seal frontier.
+/// - TODO: key uniqueness invariants?
```
Comment (ruchirK) [inline src/persist/src/indexed/encoding.rs:120]: i think this todo is a bit state -- we already enforce that all blob keys are globally unique across all arrangements in `BlobMeta`

**Review by ruchirK — APPROVED** [review]
LGTM! First 4 commits and last commit definitely look good -- id love to do another quick pass over the `ArrangementMeta` commit after lunch but it seems totally reasonable. feel free to force push!

**Comment by ruchirK** [comment]
> This to me is a non-goal. I have a reasonably strong opinion that the unit of atomic work is a PR, not a commit.

I think I agree with that, with the one exception that commits that crash on startup could make git bisection difficult (maybe that difficulty is already baked in with how we handle old versions of meta idk?)

**Comment by danhhz** [comment]
> I think I agree with that, with the one exception that commits that crash on startup could make git bisection difficult unless theres a way to git bisect across pr boundaries (maybe that difficulty is already baked in with how we handle old versions of meta idk?)

`git bisect` does exactly what you want with merge commits. Since we use either a merge commit or a squash for all our PRs, it does this now! One thing git got totally completely right

**Comment by danhhz** [comment]
I just went and did this rebase, but didn't push it because I think I've changed my mind. Thoughts on the following?

- Bumping this in every commit makes it a big hassle if I have to rebase this to merge it. For example, we might want to hold up #9239 on top of this merging, which is a practice I'd like to keep to a minimum
- Bumping it at every commit churns the golden_test.json file, which needlessly adds to the size of the git repo
- Being at a high version number might create an impression of persist volatility
- We have a limited number of versions before this has to get expanded to more than 1 byte. I've seen cases where the original author thought they'd never run out and then something unforeseen meant they did. It seems unlikely that this would be the case for us here and that we'd ever run out of 256 of them given that we're hoping to make this actually backward compatible soon, but I'm still somewhat uncomfortable with using them recklessly

**Comment by ruchirK** [comment]
Sounds like we should leave it as is. The reasons for changing were never super strong, and I think I'm most convinced by the fact that the version bumps we have available are a finite resource and it would be genuinely annoying to exhaust that.

**Comment by danhhz** [comment]
Did you get a chance to look at the ArrangementMeta commit to your satisfaction?

**Comment by ruchirK** [comment]
Yup I did! sorry should have mentioned that earlier

**Comment by danhhz** [comment]
No worries! TFTR!

File: src/persist/src/indexed/encoding.rs (hunk 1 above)
Comment (danhhz) [reply src/persist/src/indexed/encoding.rs:120]: :facepalm: Totally blanked on this, I'll fix it in the next PR.

**Comment by danhhz** [comment]
Doh! I was supposed to hold this until after the 0.11.0 release got cut but totally brain farted on it. Opening up a revert now

**Pull request merged**